:100644 100644 0000000 0000000 M	motion.conf.orig

diff --git a/motion.conf.orig b/motion.conf
index a03faf0..8101ff5 100644
--- a/motion.conf.orig
+++ b/motion.conf
@@ -77,12 +77,12 @@ flip_axis none
 
 # Image width (pixels). Valid range: Camera dependent, default: 320
-width 320
+width 640
 
 # Image height (pixels). Valid range: Camera dependent, default: 240
-height 240
+height 480
 
 # Maximum number of frames to be captured per second.
 # Valid range: 2-100. Default: 100 (almost no limit).
-framerate 2
+framerate 20
 
 # Minimum time in seconds between capturing picture frames from the camera.
@@ -119,5 +119,5 @@ rtsp_uses_tcp on
 # Name of camera to use if you are using a camera accessed through OpenMax/MMAL
 # Default: Not defined
-; mmalcam_name vc.ril.camera
+mmalcam_name vc.ril.camera
 
 # Camera control parameters (see raspivid/raspistill tool documentation)
@@ -223,5 +223,5 @@ pre_capture 0
 
 # Number of frames to capture after motion is no longer detected (default: 0)
-post_capture 0
+post_capture 5
 
 # Event Gap is the seconds of no motion detection that triggers the end of an event.
@@ -235,5 +235,5 @@ event_gap 60
 # Maximum length in seconds of a movie
 # When value is exceeded a new movie file is created. (Default: 0 = infinite)
-max_movie_time 0
+max_movie_time 30
 
 # Always save images even if there was no motion (default: off)
@@ -251,5 +251,5 @@ emulate_motion off
 # Picture with motion nearest center of picture is saved when set to 'center'.
 # Can be used as preview shot for the corresponding movie.
-output_pictures off
+output_pictures best
 
 # Output pictures with only the pixels moving object (ghost images) (default: off)
@@ -257,5 +257,5 @@ output_debug_pictures off
 
 # The quality (in percent) to be used by the jpeg and webp compression (default: 75)
-quality 75
+quality 80
 
 # Type of output images
@@ -266,5 +266,5 @@ picture_type jpeg
 
 # Use ffmpeg to encode videos of motion (default: off)
-ffmpeg_output_movies on
+ffmpeg_output_movies off
 
 # Use ffmpeg to make videos showing the moving pixels (ghost images) (default: off)
@@ -344,5 +344,5 @@ snapshot_interval 0
 # Valid values: on, off, preview (default: off)
 # Set to 'preview' will only draw a box in preview_shot pictures.
-locate_motion_mode off
+locate_motion_mode preview
 
 # Set the look and style of the locate box if enabled.
@@ -352,5 +352,5 @@ locate_motion_mode off
 # Set to 'cross' will draw a little cross to mark center.
 # Set to 'redcross' will draw a little red cross to mark center.
-locate_motion_style box
+locate_motion_style redbox
 
 # Draws the timestamp using same options as C function strftime(3)
@@ -367,5 +367,5 @@ text_right %Y-%m-%d\n%T-%q
 # Will normally be set to off except when you setup and adjust the motion settings
 # Text is placed in upper right corner
-text_changes off
+text_changes on
 
 # This option defines the value of the special event conversion specifier %C
@@ -375,5 +375,5 @@ text_changes off
 # The idea is that %C can be used filenames and text_left/right for creating
 # a unique identifier for each event.
-text_event %Y%m%d%H%M%S
+text_event %Y/%m/%d-%H:%M:%S
 
 # Draw characters at twice normal size on images. (default: off)
@@ -402,5 +402,5 @@ text_double off
 # Target base directory for pictures and films
 # Recommended to use absolute path. (Default: current working directory)
-#target_dir /tmp/motion
+target_dir /home/pi/motion/detected
 
 # File path for snapshots (jpeg, ppm or webp) relative to target_dir
@@ -411,5 +411,5 @@ text_double off
 # Note: A symbolic link called lastsnap.jpg created in the target_dir will always
 # point to the latest snapshot, unless snapshot_filename is exactly 'lastsnap'
-snapshot_filename %v-%Y%m%d%H%M%S-snapshot
+snapshot_filename %v-%Y-%m-%d_%H%M%S-snapshot
 
 # File path for motion triggered images (jpeg, ppm or .webp) relative to target_dir
@@ -420,15 +420,15 @@ snapshot_filename %v-%Y%m%d%H%M%S-snapshot
 # Set to 'preview' together with best-preview feature enables special naming
 # convention for preview shots. See motion guide for details
-picture_filename %v-%Y%m%d%H%M%S-%q
+picture_filename %v-%Y-%m-%d_%H%M%S-%q
 
 # File path for motion triggered ffmpeg films (movies) relative to target_dir
 # Default: %v-%Y%m%d%H%M%S
 # File extensions(.mpg .avi) are automatically added so do not include them
-movie_filename %v-%Y%m%d%H%M%S
+movie_filename %v-%Y-%m-%d_%H%M%S
 
 # File path for timelapse movies relative to target_dir
 # Default: %Y%m%d-timelapse
 # File extensions(.mpg .avi) are automatically added so do not include them
-timelapse_filename %Y%m%d-timelapse
+timelapse_filename %Y-%m-%d-timelapse
 
 ############################################################
@@ -453,11 +453,11 @@ stream_quality 50
 # Output frames at 1 fps when no motion is detected and increase to the
 # rate given by stream_maxrate when motion is detected (default: off)
-stream_motion off
+stream_motion on
 
 # Maximum framerate for stream streams (default: 1)
-stream_maxrate 1
+stream_maxrate 10
 
 # Restrict stream connections to localhost only (default: on)
-stream_localhost on
+stream_localhost off
 
 # Limits the number of images per connection (default: 0 = unlimited)
@@ -493,5 +493,5 @@ webcontrol_port 8080
 
 # Restrict control connections to localhost only (default: on)
-webcontrol_localhost on
+webcontrol_localhost off
 
 # Output for http server, select off to choose raw text plain (default: on)
@@ -607,5 +607,5 @@ quiet on
 # Command to be executed when a picture (.ppm|.jpg) is saved (default: none)
 # To give the filename as an argument to a command append it with %f
-; on_picture_save value
+on_picture_save /home/pi/motion/send_detection.py %f
 
 # Command to be executed when a motion frame is detected (default: none)
